server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -642,52 +652,61 @@ protected void doRun() {
-                            final String indexName = concreteIndices.getConcreteIndex(request.index()).getName();
-                            final DocWriteRequest<?> docWriteRequest = request.request();
-                            final BulkItemResponse bulkItemResponse = new BulkItemResponse(
-                                request.id(),
-                                docWriteRequest.opType(),
-                                new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e)
-                            );
+                        @Override
+                        public void onFailure(Exception e) {
+                            // create failures for all relevant requests
+                            for (BulkItemRequest request : requests) {
+                                final String indexName = concreteIndices.getConcreteIndex(request.index()).getName();
+                                final DocWriteRequest<?> docWriteRequest = request.request();
+                                final BulkItemResponse bulkItemResponse = new BulkItemResponse(
+                                    request.id(),
+                                    docWriteRequest.opType(),
+                                    new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e)
+                                );
+
+                                docStatusStats.inc(bulkItemResponse.status());
+                                responses.set(request.id(), bulkItemResponse);
+                            }
 
-                            docStatusStats.inc(bulkItemResponse.status());
-                            responses.set(request.id(), bulkItemResponse);
+                            if (counter.decrementAndGet() == 0) {
+                                finishHim();
+                            }
                         }
 
-                        if (counter.decrementAndGet() == 0) {
-                            finishHim();
+                        private void finishHim() {
+                            indicesService.addDocStatusStats(docStatusStats);
+                            listener.onResponse(
+                                new BulkResponse(responses.toArray(new BulkItemResponse[responses.length()]), buildTookInMillis(startTimeNanos))
+                            );
                         }
-                    }
-
-                    private void finishHim() {
-                        indicesService.addDocStatusStats(docStatusStats);
-                        listener.onResponse(
-                            new BulkResponse(responses.toArray(new BulkItemResponse[responses.length()]), buildTookInMillis(startTimeNanos))
-                        );
-                    }
-                }, releasable::close));
+                    }, releasable::close),
+                    span,
+                    tracer
+                );
+                try(SpanScope spanScope = tracer.withSpanInScope(span)) {
```

Comment:
Should we create the TraceableActionListener inside the try-with resources block?

Reply:
Ack. Will update it.